인가(Authorization) 구현하기

eunhasoo opened the pull request:
안녕하세요~ 
강의 내용 따라서 일단 만들어봤는데 스프링 시큐리티 구조를 잘 모르겠네요ㅠㅠ 
공부하면서 아래 내용들 추가해나가겠습니다

- Authentication과 Authority 수정
- 회원 비밀번호 인코딩 추가
- 기존 테스트 mock 제거 및 수정

한주간 잘 부탁드립니다!

eunhasoo committed:
Spring Security 의존성 추가

eunhasoo committed:
토큰 유효성 검사 메소드 추가

- 회원 인증을 수행할 필터에서 헤더값으로 들어온 문자열을 토큰으로 파싱할 수 있는지 확인할 수 있도록 유효성 검사 메소드와 테스트를 추가했습니다.
- InvalidTokenException 생성시 토큰이 아닌 메시지를 받도록 변경했습니다.

eunhasoo committed:
인증 서비스 코드 수정

- 주석을 추가했습니다.
- login 메소드에서 map을 사용하도록 변경했습니다.
- parseToken에서 발생할 예외를 체크하는 코드를 추가했습니다.

eunhasoo committed:
회원 인증 헤더 및 인증 회원 객체 추가

eunhasoo committed:
스프링 시큐리티 설정 추가

- 기존 인터셉터 사용한 방식 코드 제거

eunhasoo committed:
상품 컨트롤러 시큐리티 `@PreAuthorize` 적용

eunhasoo committed:
수행 결과와 맞지 않는 테스트 메소드명 수정

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -18,19 +19,40 @@ public AuthenticationService(UserRepository userRepository,
         Claims claims = jwtUtil.decode(accessToken);
-        return claims.get("userId", Long.class);
+        try {
+            Long userId = claims.get("userId", Long.class);
+            if (userId == null) {
+                throw new InvalidTokenException("토큰의 클레임 값이 비어있습니다.");
+            }
+            return userId;
```
Optional 을 이용해보면 어떨까요?

eunhasoo replied:
`Optional<Long>`을 반환하라는 말씀이셨을까요? 우선은 메소드 체이닝으로 아래와 같이 null을 체크하지 않고 반환하도록 변경했습니다

```java
    return jwtUtil.decode(accessToken)
            .get("userId", Long.class);
```

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,63 @@
+    protected void doFilterInternal(HttpServletRequest request,
+                                    HttpServletResponse response,
+                                    FilterChain chain) throws IOException, ServletException {
+        String accessToken = resolveToken(request);
+
+        if (StringUtils.hasText(accessToken) && jwtUtil.validateToken(accessToken)) {
```
전 아주 잘 구현하셨다고 생각해요! 

한가지 아쉬운 것은 지속적으로 `hasText` 가 계속 나온다는거에요!
`resolveToken` 이후에도 null 체크를 또 해야하는 상황이 일어나고

`userId` 또한 null 체크를 하지 않아야 할까? 라는 생각을 잠시했지만 내부에서 null 체크를 하고 있군요!

null 체크가 전파되는 느낌이 듭니다 한번 Optional을 이용해서 조금은 개선이 될 수 있지 않을까 라고 기대해봅니다!

eunhasoo replied:
null을 세밀하게 체크하려다보니 역효과로 체크하는 코드가 반복됐네요ㅠㅠ
if문을 줄여보려고 영환님 말씀대로 Optional을 이용하도록 변경해봤는데 작성한 구조가 괜찮은 방법인지 잘 모르겠습니다
혹시 Optional을 이용하는 더 좋은 구조나 팁이 있으시면 조언 부탁드립니다!

ssisksl77 replied:
확실히 Optional을 쓰다보면 뭔가 부족하다는 느낌이 듭니다! 
역시 함수형프로그래밍에서 사용하는 방식을 차용하는 것도 방법이에요!
그중 괜찮은 개념이 Either 라는 것입니다!

https://www.baeldung.com/vavr-either

eunhasoo replied:
감사합니다!! 참고하겠습니다~

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -18,19 +19,40 @@ public AuthenticationService(UserRepository userRepository,
+        return userRepository.findByEmail(email)
+                .map(user -> {
+                    if (!user.authenticate(password)) {
+                        throw new LoginFailException(email);
```
```suggestion
 return null;
```

이렇게 하면 예외를 던지는 부분이 한곳으로 일관되지않을까 싶습니다! 동일한 예외를 던지는 것이 목적이라면요!

eunhasoo replied:
넵 동일한 예외를 던지도록 하는거였는데 수정된 코드가 일관적이고 좋은 것 같아요! 반영했습니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private static List<GrantedAuthority> authorities() {
+        List<GrantedAuthority> authorities = new ArrayList<>();
+        authorities.add(new SimpleGrantedAuthority("USER"));
+        return authorities;
```
```suggestion
        return List.of(new SimpleGrantedAuthority("USER"));
```

이렇게도 가능할 것 같군요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class UserAuthentication extends AbstractAuthenticationToken {
```
주석이 있으면 좋겠습니다!

eunhasoo committed:
Optional을 사용해서 리팩토링

- 메소드 리턴값 사용시 null 체크와 if문을 최소화하도록 수정했습니다.
- UserAuthentication 클래스 주석을 추가했습니다.

eunhasoo committed:
상품 컨트롤러 테스트 리팩토링

- 클래스명을 ProductControllerAuthTest로 변경하고 인증을 테스트하도록 변경했습니다.
- MockBean를 사용하지 않도록 변경했습니다.

eunhasoo committed:
세션 컨트롤러 테스트 리팩토링

- MockBean을 사용하지 않도록 수정했습니다.
- SessionRequestData 생성자를 추가했습니다.

eunhasoo committed:
인증 서비스 테스트 Mock 사용하지 않도록 리팩토링

eunhasoo committed:
권한 리스트 List.of 사용해서 리팩토링

eunhasoo committed:
회원 생성시 비밀번호 인코딩 적용

- PasswordEncoder 빈을 추가했습니다.
- 회원 상태 및 비밀번호 검사 위치를 User 도메인에서 인증 서비스로 변경하고, 사용하지 않는 authenticate 메소드를 제거했습니다.
- 기존 테스트에서 사용자 저장시 비밀번호를 인코딩하도록 변경했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,67 @@
+                    }
+                    return null;
+                })
+                .ifPresent(userId -> {
+                    SecurityContext context = SecurityContextHolder.getContext();
+                    UserAuthentication userAuthentication = new UserAuthentication(userId);
+                    context.setAuthentication(userAuthentication);
+                });
```
아주 좋아요! 그런데 한번 쓰이고 버려지는 변수가 있군요. 객체 생성하는 메소드만으로 이해가 된다면 굳이 변수지정을 할 필요는 없어보여요! 아래 두가지 예시를 적어보았어요!

```suggestion
                .ifPresent(userId -> {
                    SecurityContextHolder.getContext()
                        .setAuthentication(new UserAuthentication(userId));
                });
```

혹은
```suggestion
                .map(UserAuthentication::new)
                .ifPresent(userAuthentication -> SecurityContextHolder
                               .getContext();
                               .setAuthentication(userAuthentication));
```

eunhasoo replied:
아래 예시 참고해서 적용했습니다! 감사합니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private Optional<String> resolveToken(HttpServletRequest request) {
+        String authorization = request.getHeader(AUTHORIZATION_HEADER);
+
+        if (StringUtils.hasText(authorization) && authorization.startsWith(BEARER)) {
+            return Optional.of(authorization.substring(BEARER.length()));
+        }
+
+        return Optional.empty();
```
이왕 옵셔널의 끝을 보겠다면 이것도 map으로 체이닝이 가능할 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/utils/JwtUtil.java:
```diff
@@ -37,7 +54,34 @@ public Claims decode(String token) {
+                .parseClaimsJws(token);
+
+            return true;
+        } catch (SecurityException
+                 | MalformedJwtException
+                 | ExpiredJwtException
+                 | UnsupportedJwtException
+                 | IllegalArgumentException e) {
```
좋아요! 혹시 다른 RuntimeException은 왜 잡지 않으시는지 알고 싶어요!

eunhasoo replied:
파싱 과정에서 문제가 있으면 곧 유효성 검사에 실패한 경우이기 때문에 `parseClaimsJws` 메소드에서 발생하는 예외만 잡아서 처리했습니다. 다른 런타임 에러가 발생하는 경우는 파싱과 관련한 예외가 아닐거라고 생각했습니다!

ssisksl77 replied:
아하! 좋습니다. 
한가지 생각해봐야할 점이 있군요!

IllegalArgumentException은 아주 범용적인 예외입니다.
원하시는 것 외에 다른 예외도 같이 잡힐 수도 있겠어요!

eunhasoo replied:
그렇네요. 메소드 주석을 다시 살펴보니 파라미터가 빈값이면 던진다고 합니다!

```java
* @throws IllegalArgumentException if the string is null or empty or only whitespace
```

이미 위에서 `StringUtils.hasText`로 검사했으니까 try-catch 블록 내부에서 굳이 잡아주지 않아도 될 것 같아서 IllegalArgumentException은 제거했어요.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -1,36 +1,56 @@
-
-        return jwtUtil.encode(1L);
+        return userRepository.findByEmail(email)
+                .map(user -> {
+                    if (!user.isDeleted() && passwordEncoder.matches(password, user.getPassword())) {
+                        return jwtUtil.encode(user.getId());
+                    }
+                    return null;
```
return null 을 하는 것이 불편할 수 있습니다! 이럴 때는 Optional의 flatMap 메소드를 사용해보세요!

eunhasoo replied:
null 대신 Optional.empty()를 반환할 수 있네요!

eunhasoo committed:
Optional 적용부 리팩토링

메소드 체이닝과 flatmap을 적용하고, 한번만 사용되는 변수는 선언없이 사용하도록 변경했습니다.

eunhasoo committed:
회원 수정시 비밀번호 인코딩하도록 수정

eunhasoo committed:
회원 수정시 본인만 요청 가능하도록 변경

- 요청에 필요한 권한이 없으면 응답을 처리할 JwtAccessDeniedHandler를 추가했습니다.
- 회원 컨트롤러에서 수정시 권한을 확인하고 없으면 AccessDeniedException을 던지도록 변경했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -1,36 +1,59 @@
+                            && passwordEncoder.matches(password, user.getPassword())) {
+                        return Optional.of(jwtUtil.encode(user.getId()));
+                    }
+                    return Optional.empty();
```
좋아요! 마지막이에요!
Optional.filter 를 사용해보세요! 

이제 flatmap의 진가를 알게될 차례입니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,66 @@
+                                    FilterChain chain) throws IOException, ServletException {
+        resolveToken(request)
+                .map(accessToken -> {
+                    if (jwtUtil.validateToken(accessToken)) {
+                        return authenticationService.parseToken(accessToken);
+                    }
+                    return null;
+                })
```
이곳도 flatMap과 filter를 사용해보세요!

eunhasoo committed:
회원 컨트롤러 테스트 작성

eunhasoo committed:
컨트롤러 메소드 파라미터로 Authentication을 받도록 수정

- 리소스 소유 확인을 UserAuthentication 클래스에서 담당하도록 변경했습니다.
- JWT 검증시 IllegalArgumentException을 잡지 않도록 했습니다.

eunhasoo committed:
가독성을 위해 filter, map 적용

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -33,13 +31,9 @@ public AuthenticationService(UserRepository userRepository,
-                    }
-                    return Optional.empty();
-                }).orElseThrow(() -> new LoginFailException(email));
+                .filter(user -> !user.isDeleted() && passwordEncoder.matches(password, user.getPassword()))
```
너무 쪼개는 거 아냐? 싶을 수 있겠지만 이곳도 두개의 filter 로 나눌 수 있을 것 같은 느낌이 들어요! 
맞을까요?

eunhasoo replied:
맞습니다!! 분리하니까 가독성이 더 좋아진거같아요 ㅎㅎ

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -47,17 +56,153 @@ void clear() {
+        @Nested
+        @DisplayName("존재하지 않는 회원 정보가 주어지면")
+        class Context_with_not_exist_user {
+            private Long id;
```
이름에 좀 더 의미를 부여해도 좋을 것 같아요!

ssisksl77 commented:
ProductService 의 역할이 너무 많은 것 같습니다!

객체지향 5원리라고 불리는 SOLID를 참고해서 분리해보아요!

eunhasoo committed:
if문 filter로 나누기

eunhasoo committed:
context 변수명 명확하게 수정

eunhasoo committed:
ProductService 역할 분리

- 상품 조회 역할은 ProductQueryService에서 맡도록 변경하고 테스트를 추가했습니다.
- 상품 변경 역할은 productCommandService에서 처리하도록 변경하고 테스트를 추가했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/ProductCommandService.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * 상품 변경 서비스.
+ */
+public interface ProductCommandService {
```
아주 좋아요! 좀 더 극단적으로 해보자면 더 쪼갤 수 있지 않을까요?!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/ProductQueryServiceTest.java:
```diff
@@ -0,0 +1,120 @@
+            @DisplayName("해당 아이디의 상품을 반환한다")
+            void it_returns_all_products() {
+                Product product = productQueryService.get(productId);
+                assertThat(product).isNotNull();
```
"해당 아이디의 상품을 반환한다" 이긴 하지만 체크하는 것은 null 여부네요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -43,10 +43,10 @@ public class Product {
-        this.maker = source.maker;
-        this.price = source.price;
-        this.imageUrl = source.imageUrl;
+    public void update(String name, String maker, int price, String imageUrl) {
```
아주좋아요! 
굳이 트집을 잡자면!

`update(String, String, int, String)` 의 경우 같은 타입이 3개나 있어요! 혹시 실수를 막기위한 방법으로 어떤 옵션이 있을지 공유해보아요!

eunhasoo replied:
1. 변경전의 기존 코드처럼 update를 위한 객체를 만들어서 사용한다. 도메인이 DTO에 종속될 수 있다는 단점이 있다.
2. update에 Builder 패턴을 적용한다. 불필요한 새로운 인스턴스를 생성하게 되는 단점이 있다.

2를 기반으로한 [이 방법](https://stackoverflow.com/a/59967252)도 괜찮은 거 같아요. 그런데 생성후 변경하면 안되는 필드를 변경으로부터 어떻게 막을 수 있을지는 고민이 필요할 것 같아요!

ssisksl77 replied:
- 빌더도 좋습니다아니면 
- 새로운 객체를 만들 수도 있죠!
- 혹은 Command 패턴을 고려해볼 수도 있겠습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/ProductCommandServiceTest.java:
```diff
@@ -0,0 +1,166 @@
+        @DisplayName("수정할 상품 아이디와 입력값이 주어지면")
+        class Context_with_valid_input {
+            private Long productId;
+            private final ProductData updateProductData = updateData();
```
현재로도 좋습니다.

한가지 아까운 점은 createData와 updateData가 당연히! 다르다는 것을 알지만서도!

혹여나 createData의 정보가 updateData와 동일한 값으로 바뀔 수도 있어보여요!

동일한 값으로 바뀌면 테스트가 의미없이 성공하는 것이 되겠죠!

eunhasoo replied:
`동일한 값으로 바뀔수 있다`는 말이 어떤 의미일까요? 🥹
생성과 수정에 각각 다른 DTO가 있으면 좋겠다는 말씀이실까요? 
일단은 createData()의 정보를 기반으로 수정한 내용을 updateData()가 갖도록 변경해보았습니다.

영환님 덕에 Optional API에 조금 더 친숙해질 수 있었네요! 그동안 고생 많으셨습니다!!!

ssisksl77 replied:
만약에. `updateData` 내부구현을 `createData` 와 동일해졌다 하더라도. 테스트는 전부 성공할 것이라는 겁니다!
그것도 당연히 성공해야하는 테스트지만! 테스트가 유효하지 않게 될 수 있다는 것이죠!

음... 요 링크를 한번 보시면 좋겠어요!

https://enterprisecraftsmanship.com/posts/dry-damp-unit-tests/

ssisksl77 replied:
감사합니다. `flatMap` 의 강력함을 공유드리지 못한 점이 아쉽지만! 그래도 좋은 경험이 되었다니 기분이 좋네요!

eunhasoo committed:
테스트 의도에 맞도록 검증문 수정

eunhasoo committed:
생성 입력값을 기반으로 수정 입력값을 받을 수 있도록 테스트 수정

ssisksl77 commented:
한주간 고생많으셨습니다!
